blog/content/post/2025-07-xx-edk2-on-odroid-m2.md

```diff
@@ -0,0 +1,325 @@
+mass, I've had to set a low PWM period and duty (50 microseconds). This helps
+ensure that the fan doesn't make a lot of noise at lower PWM duty cycles.
+
+Here's the code snippet.
+
+```c
+PWM_DATA pwm_data = {
+    .ControllerID = PWM_CONTROLLER0,
+    .ChannelID = PWM_CHANNEL0,
+    .PeriodNs = 50000,
+    .DutyNs = 50000,
+    .Polarity = FALSE,
+}; // PWM0_CH0
+
+VOID
+    EFIAPI
+    PwmFanIoSetup(
+        VOID)
+{
+  GpioPinSetFunction(1, GPIO_PIN_PA2, 0xB); // PWM0_M2
+  RkPwmSetConfig(&pwm_data);
+  RkPwmEnable(&pwm_data);
+}
+```
+
+## Building
+
+To build the board, simply execute the build.sh script from the root of the
+edk2-rk3588 repo:
+
+```bash
+./build.sh --device odroid-m2
+```
+
+This produces a binary that we can flash to hardware:
+
+```
+~/Development/Dasharo/edk2-rk3588
+ => FIT build done
+ => Building 8MB NOR FLASH IMAGE
+34+0 records in
+34+0 records out
+17408 bytes (17 kB, 17 KiB) copied, 0.000452301 s, 38.5 MB/s
+300+0 records in
+300+0 records out
+307200 bytes (307 kB, 300 KiB) copied, 0.000979615 s, 314 MB/s
+5745+1 records in
+5745+1 records out
+5883392 bytes (5.9 MB, 5.6 MiB) copied, 0.012152 s, 484 MB/s
+Build done: RK3588_NOR_FLASH.img
+```
+
+## Flashing
+
+The Odroid can boot from microSD cards, so let's write the image to a card:
+
+```bash
+sudo dd if=RK3588_NOR_FLASH.img of=/dev/sdb
+```
+
+Now, let's insert the card into the SBC and power it on.
+
+![img](/img/odroid_edk2.jpg)
+
+Success!
+
+## Summary
+
+While the port is still incomplete and many things aren't working as expected
+yet (booting Windows, USB3 storage in firmware, LEDs etc), the port is already
+capable of installing and booting Ubuntu 25.04 with most things just working.
+We are still far from a fully-fledged Dasharo release for an Arm-based platform,
+but this experiment gives us some experience and brings us that much closer to
+supporting this architecture.
+
+The pull request to edk2-rk3588 adding the Odroid M2 port is available
+[here](https://github.com/edk2-porting/edk2-rk3588/pull/218).
+
+
+Huge thanks to the edk2-porting people for maintaining the edk2-rk3588
+repository and providing a solid base to build on. Check out their repository:
+[link](https://github.com/edk2-porting/edk2-rk3588).
+
+Unlock the full potential of your hardware and secure your firmware with the
+experts at 3mdeb! If you're looking to boost your product's performance and
+protect it from potential security threats, our team is here to help. [Schedule
+a call with
+us](https://cloud.3mdeb.com/index.php/apps/calendar/appointment/n7T65toSaD9t) or
+drop us an email at `contact<at>3mdeb<dot>com` to start unlocking the hidden
+benefits of your hardware. And if you want to stay up-to-date on all things
+firmware security and optimization, be sure to [sign up for our
+newsletter](https://3mdeb.com/subscribe/3mdeb_newsletter.html). Don't let your
+hardware hold you back, work with 3mdeb to achieve more!
```

Comment:
We need a CfP fothe the upcoming DUG/vPub: https://cfp.3mdeb.com/developers-vpub-0x10-2025/
Main reason for this blog post is event promotion.

Reply:
@mkopec not only that but also submission of your talk.